### ISSUE TITLE
Develop Django command for embedding batch work

### ISSUE THREAD

**Issue by albertisfu** [freelawproject/courtlistener#4559]
This task consists of implementing the Django command to perform the initial batch work. It basically consists of two parts:

- A command that pulls up Opinion texts from the database and creates a batch of opinion texts to send to the microservice.

This is related to your comment:

> I'd guess this is less about how many opinions to do at once and more about how long those opinions are.
> 

I'm thinking that the command can only iterate over all the Opinion pks in DB so we can just send a chunk of pks to the Celery task that will have to do the work:

- Request opinion texts from the pks passed.
- Create the request body for these opinion texts
- Send the request to the microservice
- Wait for the microservice response
- Store embeddings into S3

In that way, the task would need to only hold opinion pks. However, the disadvantage is that some requests can be just a few KBs while others can be many MBs.

- The alternative is to request the opinion texts within the command and set a batch threshold that we determine performs better to generate embeddings in the microservice, say 1MB or 10MB, wherever we determine. So we extract opinion texts and when we're close to this limit, we pass the batch of texts to the task.
    - The disadvantage is that tasks will require holding opinion texts instead of retrieving them within the task.
    - But probably this is OK since we don't expect to have many tasks in the queue. Considering we'd just set up an equivalent number of Celery workers as we have microservice instances. If we use queue throttling, the queue should be small all the time.

**The output of this issue would be:**

- A PR that includes the Django command to request opinion texts in batches.

**Some questions:**

@legaltextai, in terms of performance and efficient use of resources on our GPU/CPU machines used for embedding work, would it affect performance to have requests with varying amounts of text to embed? For example, sometimes we send 1MB for embedding, and other times 100KB. Is it more efficient to always request a fixed amount of text for embedding?

If so, is it possible to determine the ideal size of text to request in a single embedding request?

**Comment by elisa-a-v** [comment]
To the best of my understanding, the main goal of this particular issue will be to store the text embeddings generated by the new microservice into S3 for all Opinions we have. Then, in a later issue, we'll use those to index the embeddings in ES.

I talked with both @albertisfu and @ERosendo and we've come up with two alternatives to achieve this:

### Django command + Celery task that does most of the work

This approach would involve writing one command that iterates over Opinions and builds batches that can then be sent to a Celery task to be processed. This task would then send a batch request to the microservice which would return the embeddings, and then send those embeddings to S3. The task would handle a single batch per execution, so we'd have to add as many tasks to the Celery queue as the number of batches to process.

This alternative would be limited by the number of Celery workers, and it would use up some of the server's resources, but it would leverage the retry mechanisms we have in place for Celery tasks.

### Lambda function

For this approach we would need to create manifests that the Lambda function can then consume to do the batch work of calling the microservice and storing the embeddings in S3.

These manifests would be created and sent to S3 by a Django command that we already have (`make_aws_manifest_files`), and they would contain rows with ranges of Opinion ids, each row representing a batch that the Lambda function would process at a time.

We should use multiple manifests because the number of Opinions to process is huge (several millions), so we want the lambda function to execute as fast a possible in case something happens and we need to correct it.

### Questions and considerations

1. How many concurrent requests can the microservice handle?
	- If this number is really small, the Lambda function might not be worth it, as it shines when it handles multiple concurrent operations—as I understand it, our current setup has ~11 Celery workers vs the ~1000 concurrent operations supported by the Lambda function—but I think this can be configured, so I'm not entirely sure about this particular point.
2. If the batch size affects microservice performance, we should check the distribution of opinion sizes to determine the optimal number of opinions per batch to send to the microservice.
	- If the standard deviation for these is rrrreally high, we might need to opt for a different approach to build batches instead of a fixed number of opinions per batch.
	- If the microservice isn't affected by the batch size then this would be irrelevant and we can stick to the fixed number of Opinions per batch.
3. Is the microservice already deployed and ready to use? I don't think this would affect this discussion too much, but it could enable some testing.

---

I was initially more inclined to go the Django command + Celery task alternative mostly because I'm more familiar with it and it seemed like we already had a full infrastructure in place that would take care of things like retries and throttling, but after talking to @ERosendo, I now know we also have a bunch of tools to use Lambda functions for batch work, and it *seems* like it would actually be less work to develop this by leveraging them.

Alberto and Eduardo, please correct me if I got anything wrong or if I missed something else.

@mlissner what do you think?

**Comment by albertisfu** [comment]
> How many concurrent requests can the microservice handle?

Yeah, this question is key. Using a Celery task, we can control the speed (via throttling) of scheduling new tasks dynamically based on the microservice's response time, preventing it from being overwhelmed and avoiding timeouts. Is there a way to achieve the same in a Lambda function when performing concurrent operations?

**Comment by mlissner** [comment]
Elisa and I chatted about that a bunch today. I was actually suggesting that we throttle it based on the number of celery workers too, but similar idea. I don't know what lambda makes possible, but I'm pushing for not-lambda pretty hard. :)

**Comment by elisa-a-v** [comment]
Well, I was pretty convinced by Eduardo's persuasive arguments in favor of Lambda functions, but after discussing this with both Mike and Rachel today, I'm actually more inclined to go the Celery route.

The microservice would indeed be a big bottleneck to actually leverage the Lambda advantages, and it seems easier to throttle the whole process by using Celery while also being able to scale the workers as needed. The retry mechanisms we have in place for Celery tasks are also a big reason to stick to them as failures in a Lambda function would probably involve more manual work. Oh and this way we would be keeping all the code in one place, which is always nice.

Now the idea would be to do something similar to the first alternative I outlined in my previous comment, in that we would still use a Django command to build batches, and we would then send those to Celery to get the embeddings from the microservice and upload them to S3—but the difference is now we would do 2 different Celery tasks: one for requesting the embeddings from inception, and another one to upload them to S3. This approach would enable us to make the most of the GPU by keeping it busy, without having to wait for a task to finish uploading the embeddings to S3 before starting to get the embeddings for the next batch.

The steps would then be:

1. The Django command iterates over Opinions in our DB and creates batches of opinions.
	- We initially considered using either file size or page count for building batches, but after talking to Rachel, I now think the best way to do this would be using tokens. If this isn't easy I think defaulting to word count instead of page count would still be better, but I don't think tokenizing (?) the text would be too hard.
	- To do this we need to fetch from DB not just the Opinion IDs, but also their text, so we can determine the optimal batch size.
	- After we reach a certain threshold for the batch size (either in terms of word count or token count), we add a Celery task to the queue, passing all the Opinion IDs in the current batch to the task. We do NOT pass the text here, even though we will need it again to send it to inception, so as not to overload the Redis memory.
	- Rinse and repeat until we've ran out of Opinions.
2. The first Celery task gets a batch of Opinion IDs and queries the DB again to get their text. Then, it sends a batch request to inception sending all the Opinion IDs with their text.
	- I don't really like the idea of hitting the DB twice per Opinion to retrieve the same data, but if it starts affecting performance, we can always query a replica instead of the prod DB (which we can easily achieve with `QuerySet.using()`). The alternative would be to send the already fetched text as an argument in the task alongside the IDs, but we will already introduce some overhead in terms of redis memory usage in the second task, so this seems like a reasonable compromise.
3. Once the first task has the embeddings, it then sends a new task to a (different?) queue, which would take the actual embeddings in an argument to then upload to S3.
	- Big concern here was the Redis memory overhead of sending tasks with arguments bigger than simple IDs to a Celery queue, but as Rachel explained, the embeddings have a somewhat fixed size (most of them will have a single row with a fixed size, while the biggest ones will have more rows, but all rows have the exact same size), and this shouldn't be big at all, since each row in the embedding would have ~750 numbers in it. A JSON with a few lists of numbers <800 items long doesn't sound bad at all.

@mlissner would you say this sums up everything we talked today?

and @ERosendo, do you vehemently disagree or does this make sense to you? 😅

**Comment by mlissner** [comment]
Sounds spot on and simple enough to me. One loop, two tasks. You'll want to identify which arguments the command should have. I can think of:

 - `--database`: Let the user decide which DB name to use (default to `default`).
 - `--batchsize`: How big of a batch (default to something @rachlllg recommends?)
 - `--start-id`: Which opinion ID should we start with, in case it crashes? (default to 0)
 - `--vectorizing-queue`: Which celery queue to use for vectorizing
 - `--upload-queue` which celery queue for uploading to S3 (default: batch1)

And probably a couple others we normally have.

I think tokenizing just to count tokens instead of words seems unnecessary. We gain accuracy by doing so, but does that really matter? It adds complexity and a dependency to our codebase (we don't have anything to count tokens yet, I don't think).

Sounds good to me though on the whole!

**Comment by elisa-a-v** [comment]
> I think tokenizing just to count tokens instead of words seems unnecessary. We gain accuracy by doing so, but does that really matter? It adds complexity and a dependency to our codebase (we don't have anything to count tokens yet, I don't think).

Well, I found this command ([cl_compute_token_count](https://github.com/freelawproject/courtlistener/blob/main/cl/search/management/commands/cl_compute_token_count.py)) so we actually already have the logic and the dependency for it. It includes two helper functions (`get_clean_opinion_text` and `get_token_count_from_string`) that do exactly what I need, and I think I could add them as model properties in the Opinion model, then use that in the Django command. Seems simple enough?

And even if we decide not to count the tokens, I still think adding an Opinion property with the `clean_text` makes sense.

**Comment by rachlllg** [comment]
> Well, I found this command ([cl_compute_token_count](https://github.com/freelawproject/courtlistener/blob/main/cl/search/management/commands/cl_compute_token_count.py)) so we actually already have the logic and the dependency for it.

Token counts from `tiktoken` are not exactly the same as using AutoTokenizer with the specified model, but for this exercise I think they are fairly close.

> --batchsize: How big of a batch (default to something @rachlllg recommends?)

On the batch size, I think it would depend on the machine we have (how much memory and number of CUDA cores). Below is what the ModernBert paper suggests, I think the red squared parts are what we would be interested in. So on an NVIDIA RTX 4090 GPU, they suggested a max batch size of 98 documents. If our GPU is comparable to this one, may I suggest we do a bit of trial and error, start at a smaller batch size (maybe 60 documents or 250K tokens?) And gradually increase until our GPU is maximally utilized? 

(The Long Variable dataset is: "varying-length document sets, where the number of tokens are defined by a normal distribution centered on half the maximum sequence length 4096 tokens")

![Image](https://github.com/user-attachments/assets/f2312ff0-a6ce-4f67-94d1-1ae95831ef00)

**Comment by mlissner** [comment]
Cool. @elisa-a-v, this implies one other argparse argument, which is `--count` (or similar), so that the script stops processing after so many items. If we're doing trial and error, it'll be a lot easier if we can send 1,000 or 5k or however many items at a time, test performance, then go again. Otherwise, we have to kill it each time.

**Comment by rachlllg** [comment]
Another thing I just thought of, if we have opinions that are very long (say more than a few hundred pages), it may exceed the batch_size on an opinion level (causing OOM), we might want some logging for these opinions and use preprocessing to chunk them to smaller opinions first.

**Comment by elisa-a-v** [comment]
@rachlllg can we know in advance what size would cause OOM? I'm thinking that's not something we want to log after the fact, but rather we should prevent the request for those big opinions and log the ones we skipped.

**Comment by rachlllg** [comment]
Just want to put the latest Inception updates and how it will impact the embedding generation here so everyone is aware (thanks @albertisfu for the questions!)

We have changed the context window size from 8192 tokens to 512 tokens in the latest Inception update, this will allow us to highlight the section of the opinion that best matches a user query for better user experience.

Previously, with context window of 8192 tokens, most opinions will be embedded to one vector of size 768 (as most opinions are not longer than 8192 tokens).

Now, with context window of 512 tokens (which loosely translates to 200 words), most opinions will be embedded to numerous vectors of size 768. 

What this means is, with context window of 512 tokens, 250K tokens would now translate to ~488 vectors, each with size 768.  In the previous version with context window of 8192 tokens, 250K tokens would translate to ~30 vectors, each with size 768, so  ~16x increase.

Also, now that we are using the shorter context window, we should be looking at the "short" column in the below [screenshot](https://arxiv.org/pdf/2412.13663), which suggests a max batch size of 1604 batches, each with context window of 512 tokens, which is ~800K tokens, optimized for NVIDIA RTX 4090, which has 24GB GPU RAM. (Note that their experiments were run on texts with normal distribution surrounding a 256 token mean, so the max would be smaller than 800K tokens).

![Image](https://github.com/user-attachments/assets/55067dfd-f93c-46e2-98bd-e907aa76b8aa)
